allennlp/data/dataset_readers/semantic_parsing/grammar_based_text2sql.py

```diff
@@ -0,0 +1,140 @@
+        in the training set.
+    """
+    def __init__(self,
+                 schema_path: str,
+                 use_all_sql: bool = False,
+                 remove_unneeded_aliases: bool = True,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 cross_validation_split_to_exclude: int = None,
+                 lazy: bool = False) -> None:
+        super().__init__(lazy)
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+        self._use_all_sql = use_all_sql
+        self._remove_unneeded_aliases = remove_unneeded_aliases
+        self._cross_validation_split_to_exclude = str(cross_validation_split_to_exclude)
+
+        self._sql_table_context = Text2SqlTableContext(schema_path)
+        self._world = Text2SqlWorld(self._sql_table_context)
+
+    @overrides
+    def _read(self, file_path: str):
+        """
+        This dataset reader consumes the data from
+        https://github.com/jkkummerfeld/text2sql-data/tree/master/data
+        formatted using ``scripts/reformat_text2sql_data.py``.
+
+        Parameters
+        ----------
+        file_path : ``str``, required.
+            For this dataset reader, file_path can either be a path to a file `or` a
+            path to a directory containing json files. The reason for this is because
+            some of the text2sql datasets require cross validation, which means they are split
+            up into many small files, for which you only want to exclude one.
+        """
+        files = [p for p in glob.glob(file_path)
+                 if self._cross_validation_split_to_exclude not in os.path.basename(p)]
+
+        for path in files:
+            with open(cached_path(path), "r") as data_file:
+                data = json.load(data_file)
+
+            for sql_data in text2sql_utils.process_sql_data(data,
+                                                            use_all_sql=self._use_all_sql,
+                                                            remove_unneeded_aliases=self._remove_unneeded_aliases,
+                                                            # TODO(Mark): Horrible hack, remove
+                                                            schema=self._sql_table_context.schema):
+                instance = self.text_to_instance(sql_data.text, sql_data)
+                if instance is not None:
+                    yield instance
+
+    @overrides
+    def text_to_instance(self,  # type: ignore
+                         query: List[str],
+                         sql: text2sql_utils.SqlData = None) -> Instance:
+        # pylint: disable=arguments-differ
+        fields: Dict[str, Field] = {}
+        tokens = TextField([Token(t) for t in query], self._token_indexers)
+        fields["tokens"] = tokens
+
+        if sql is not None:
+            try:
+                action_sequence, all_actions = self._world.get_action_sequence_and_all_actions(sql.sql)
```

Comment:
Not for this PR necessarily, but `all_actions` here is constant across all instances, yet we're constructing the object for every instance, and making the same `ProductionRuleField` list for them.  It'd be nice to figure out a way around this at some point...  (See #1809)

Reply:
I'm not sure it actually will be the same across instances - e.g consider the case that we are given anonymized data with entity information - then we want to constrain the string productions to be only the entities with the correct type, rather than any of them. But I see what you mean I think it's also a slow part of the code, but I think it's unlikely to matter too much for the text2sql datasets as they are all reasonably sized.